samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -155,10 +161,10 @@ public void join(String version, String participantId) {
    */
   public void expire(String version) {
     String barrierStatePath = keyBuilder.getBarrierStatePath(version);
-    State barrierState = zkUtils.getZkClient().readData(barrierStatePath);
+    State barrierState = State.valueOf(zkUtils.getZkClient().readData(barrierStatePath));
```

Comment:
Are there compatibility concerns? eg: when data written using serde format is read using the string serializer? 
May not be an issue if we always do a write before we do a read. 

Either-ways, would be good to understand the behavior wrt. rolling upgrades

Reply:
In addition to switching to string serializer, I've also changed the zookeeper base path that is used to store all the coordination data in zookeeper(Refer protocol version comment).  When an application picks-up this change, the old state will not be considered for `JobModel` generation. We will not de-serialize the old state with the new serializer. 

> Compatibility concerns:

Currently there are only stateless standalone jobs in production. So the previous state is not relevant and not useful for the task redistribution on processor shuffle. 

> Rolling upgrade. 

For rolling upgrade scenario, the newer processors will be storing the coordination data to new zookeeper namespace and the older processors will be storing the coordination data to old zookeeper namespace. So there will be no serialization exceptions. Though, there could be a brief rolling-bounce window, where in some processors could be doing duplicate stream processing. I think we can document these situations in the release wiki and guide users. 

Despite this notable problem, I think this patch is the way forward. It will help in viewing the standalone zookeeper state with tools in LinkedIn and open source. Will allow us to build tools in non-java languages for displaying standalone-state(Building dashboard for standlaone).

Reply:
FYI, I don't think anyone disagrees that this patch is the way forward..